csrc/device_lower/analysis/tensor_memory.h

```diff
@@ -45,20 +48,112 @@ TensorMemoryInfo computeTMemInfo(Fusion* fusion);
 // relinquishes the right to allocate, the next CTA that is blocked will be
 // unblocked and can acquire the mutex to allocate TMem.
 //
-// Currently, our TMem allocation strategy is as naive as follows:
-// We assume there is at most one TensorView on TMem in the fusion. With this
-// assumption, we don't have to worry about where to place different tensors on
-// TMem. We will traverse the fusion to look for a TMem TensorView. If we can
-// find such a TensorView, we will generate a tcgen05.alloc and
-// tcgen05.relinquish_alloc_permit at the beginning of the kernel. We do not
-// dealloc TMem for now.
+// The tcgen05.alloc instruction is like the following:
+//   tcgen05.alloc [dest], nCols
+//
+// There are three important things to note about this instruction:
+//
+// 1. The output of this instruction is in shared memory address.
+// 2. The unit of allocation is 32 whole columns of tensor memory. And nCols
+//    must be a power of two.
+// 3. The right to allocate is like a mutex and will serialize CTA scheduling.
+//    The tcgen05.alloc is blocking when there is no space to allocate.
+//
+// The point 1 above is not a big trouble for us, but we need to make sure we
+// allocate the address tensor in shared memory before allocating the tensor
+// memory. But the point 2 and 3 can be a big challenge. There are basically
+// two things to worry about when allocating tensor memory:
+//
+// 1. Fragmentation. When the tensor does not occupy all lanes or the tensor's
+// size is not a power of two columns or < 32 columns, naively allocating all
+// lanes with 32 or higher power of 2 columns could waste some space. In a
+// perfect world, it would be nice to have a 2D allocator that is capable
+// merging the allocation of multiple tensors into a single tcgen05.alloc.
+// For example, if tv0 and tv2 both has 64 rows and 32 columns, we can allocate
+// tv0 on the first 64 lanes, and tv1 on the next 64 lanes. Another example is,
```

Comment:
Lanes here mean rows and there are 128 rows. Correct?

Reply:
Yes, and I think we use the term "lane" and "row" interchangeably: https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#tensorcore-5th-generation-family-instructions
<img width="1102" alt="image" src="https://github.com/user-attachments/assets/332032a2-25e2-4564-bf9e-42c77a680817" />